packages/base-manager/src/utils.ts

```diff
@@ -298,3 +298,70 @@ export function bufferToBase64(buffer: ArrayBuffer): string {
 export function base64ToBuffer(base64: string): ArrayBuffer {
   return toByteArray(base64).buffer;
 }
+
+/**
+ * A map that chunks the list into chunkSize pieces and evaluates chunks
+ * concurrently.
+ *
+ * @param list - The list to map over
+ * @param chunkOptions - The options for chunking and evaluating.
+ * @param fn - The function to map, with the same arguments as an array map
+ * @param thisArg - An optional thisArg for the function
+ * @returns - the equivalent of `Promise.all(list.map(fn, thisArg))`
+ */
+export async function chunkMap<T, U>(
+  list: T[],
+  chunkOptions: chunkMap.IOptions,
+  fn: (value: T, index: number, array: T[]) => Promise<U> | U,
+  thisArg?: any
+): Promise<U[]> {
+  // Default to equivalent to Promise.all(list.map(fn, thisarg))
+  const chunkSize = chunkOptions.chunkSize ?? list.length;
+  const concurrency = chunkOptions.concurrency ?? 1;
+
+  const results = new Array(list.length);
+  const chunks: (() => Promise<void>)[] = [];
+
+  // Process a single chunk and resolve to the next chunk if available
+  async function processChunk(chunk: any[], start: number): Promise<void> {
+    const chunkResult = await Promise.all(
+      chunk.map((v, i) => fn.call(thisArg, v, start + i, list))
+    );
+
+    // Splice the chunk results into the results array. We use
+    // chunkResult.length because the last chunk may not be full size.
+    results.splice(start, chunkResult.length, ...chunkResult);
+
+    // Start the next work item by processing it
+    if (chunks.length > 0) {
+      return chunks.shift()!();
+    }
+  }
+
+  // Make closures for each batch of work.
+  for (let i = 0; i < list.length; i += chunkSize) {
+    chunks.push(() => processChunk(list.slice(i, i + chunkSize), i));
+  }
+
+  // Start the first concurrent chunks. Each chunk will automatically start
+  // the next available chunk when it finishes.
+  await Promise.all(chunks.splice(0, concurrency).map(f => f()));
```

Comment:
Very intersting pattern!
This is the place where we can protect against the io rate limit I think (not tested).
```suggestion
  const concurrency = chunkOptions.rate ?? Infinity;
  const delay = (sec) => new Promise((resolve) => setTimeout(resolve, sec*1000));
  const rateLimit (f) = > Promise.all(f(), delay((concurrency * chunkSize) / rate)).then(v => v[0]);
  await Promise.all(chunks.splice(0, concurrency).map(f => rateLimit(f())), de);
```